### ISSUE TITLE
Explicit connection reuse

### ISSUE THREAD

**Issue by timrburnham** [pacman82/arrow-odbc-py#92]
It would be convenient to be able to re-use specific database connections for explicitly sequential tasks. For instance, declare a temporary table, insert some rows from an Arrow table, and join against that session table in a subsequent select.

Things I have tried unsucessfully:
- enable_odbc_connection_pooling(), as mentioned in #43
- decorate connect.py:connect_to_database() with @functools.cache()
  - (yes I knew it was a bad idea, thought it was worth a try)

I assume there are Rust ownership issues that make this difficult, but it would be amazing for ETL jobs which compute intermediate data sets. Right now I'm creating tables dynamically and dropping them after.

**Comment by pacman82** [comment]
Hello @timrburnham ,

> It would be convenient to be able to re-use specific database connections for explicitly sequential tasks. For instance, declare a temporary table, insert some rows from an Arrow table, and join against that session table in a subsequent select.

If it is about the saving the time to reconnect to the database, connection pooling does enable this. This sounds more like you want everything happening in the same transaction? Also, while declaring temporary tables is currently possible in `arrow-odbc` because it does execute arbitrary SQL statements the scope and intention of the package is currently only data insertion in and out of tables. With my current day-job I am a bit anxious if I could tackle the scope of another `pyodbc`.  I am already maintaining the `odbc-api` bindings for Rust.

So I would need to understand very precisely why you want these commands on the same connection. Maybe there is something here which still fits in the scope of the `arrow-odbc` Python bindings. It could very well be.

Alternatively you could also try:

* Using `arrow-odbc` directly from Rust. You can do everything you want in conjunction with `odbc-api`. Even fancier stuff like reusing `Connection` and `Statement` handles.
* Switching to `turbodbc` it is potentially harder to install and behaves differently in some situations, but it does allow fast bulk fetch into arrow arrays and it is broader in scope, including explicit connections in Python code.

> I assume there are Rust ownership issues that make this difficult

As I see it every piece of Software has ownership issues. It is just in Rust the compiler tells you about them.

**Comment by timrburnham** [comment]
Thank for the quick analysis! I don't think I need transaction control, just the same session--I really thought it might work with connection pooling. I'm not clear what's being reset.

In this test case, I'm using the DuckDB driver under UnixODBC, since it's simple to setup and use. For reference, my ~/.odbc.ini contains:

```ini
[DuckDB]
Driver=DuckDB Driver
Database=:memory:
```

and my ~/.odbcinst.ini:
```ini
[DuckDB Driver]
driver = /home/tim/libduckdb_odbc.so
```

Here's a non-working case, using arrow-odbc:
```python
arrow_odbc.enable_odbc_connection_pooling()

# create temp table, DuckDB syntax
sql = """\
create temporary table temp_keys (
  k varchar(128)
)
on commit preserve rows
;"""
arrow_odbc.read_arrow_batches_from_odbc(sql, 'DSN=DuckDB;')

# insert a bunch of rows from Arrow Table
filter = pa.Table.from_pydict({"k": ["key1", "key2", "key3"]})
reader = pa.RecordBatchReader.from_batches(filter.schema, filter.to_batches())
arrow_odbc.insert_into_table(
    reader=reader,
    chunk_size=1000,
    table="temp_keys",
    connection_string='DSN=DuckDB;',
)

sql = """\
select v from
  values ('key1', 1), ('key6', 6) as fake_big_table(k, v)
inner join temp_keys
  using (k)
;"""
results = arrow_odbc.read_arrow_batches_from_odbc(sql, 'DSN=DuckDB;')
```

Results:
```
>>> results = arrow_odbc.read_arrow_batches_from_odbc(sql, 'DSN=DuckDB;')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tim/venv/lib64/python3.11/site-packages/arrow_odbc/reader.py", line 493, in read_arrow_batches_from_odbc
    reader.query(
  File "/home/tim/venv/lib64/python3.11/site-packages/arrow_odbc/reader.py", line 127, in query
    raise_on_error(error)
  File "/home/tim/venv/lib64/python3.11/site-packages/arrow_odbc/error.py", line 30, in raise_on_error
    raise Error(error_out)
arrow_odbc.error.Error: ODBC emitted an error calling 'SQLExecDirect':
State: 42000, Native error: 0, Message: ODBC_DuckDB->PrepareStmt
Catalog Error: Table with name temp_keys does not exist!
Did you mean "pg_views"?
LINE 3: inner join temp_keys
                   ^
```

And after pip installing the native DuckDB driver, here's a working version where we explicitly reuse same connection:

```python
duck = duckdb.connect(':memory:')
sql = """\
create temporary table temp_keys (
  k varchar(128)
)
on commit preserve rows
;"""
duck.sql(sql)

filter = pa.Table.from_pydict({"k": ["key1", "key2", "key3"]})
duck.sql("insert into temp_keys(k) select k from filter;")

sql = """\
select *
from values ('key1', 1), ('key6', 6) as fake_big_table(k, v)
inner join temp_keys
  using (k)
;"""
results = duck.sql(sql).to_arrow_table()
```

Results:
```
>>> results
pyarrow.Table
k: string
v: int32
----
k: [["key1"]]
v: [[1]]
```

The general idea is, if I have a million rows in a table, and I only want a thousand of them, if I already know the keys it's easy to upload them into a temp table and join against my big table. You can get pretty far binding parameters to where clauses, but joins are better in some cases.

**Comment by pacman82** [comment]
Hello @timrburnham ,

thanks for the detailed response and sorry for the delayed answer. The use case is legit, however it is quite removed from there `arrow-odbc` is currently in terms of interface. Your first statement already illustrates this well:

```
# create temp table, DuckDB syntax
sql = """\
create temporary table temp_keys (
  k varchar(128)
)
on commit preserve rows
;"""
arrow_odbc.read_arrow_batches_from_odbc(sql, 'DSN=DuckDB;')
```

Although working, this is of course not the intended use of `read_arrow_batches_from`, as you are only interessted in the "side effect" of temporary table creation.

I'll give it some thought. However please not that, at least during the month of May, I won't have any time to act on this, so please be aware that you will need a workaround / different solution in the meanwhile.

Best, Markus

**Comment by timrburnham** [comment]
Thanks very much, Markus! Using permanent tables is fine as a workaround, of course, I'm only thinking of the ergonomics. I really really like using arrow-odbc, thanks for a fantastic package!

**Comment by pacman82** [comment]
This is mostly a note to myself:

I have not decided yet, whether to take this into the scope of the `arrow-odbc` Python bindings. If so however I can see two designs working.

1. `Connection`s are explicitly instantiated by the user. They have their own representation in Python code and could be passed explicitly in the creation of readers and writers. On the Rust side this would imply storing them in an `Arc<Mutext<_>>` instead of taking full ownership of them. `Mutex` would be required explicitly to ensure error messages end up correctly on the thread that caused them. This design would not only allow to use the same connection for multiple statements in succession, but multiple statements might be active at the same time utilizing the same connection. This is supported according to ODBC, but I am fearful this would create hard to debug errors with some drivers.

2. `Connection`s are not represented explicitly in user code. Instead there is only one handle to an object representing all the possible states of `arrow-odbc`. On the Rust side this would imply unifying the `Reader` and `Writer` into a single `enum`. In additon we must not close the `Connection` as soon as the `Reader` is consumed, but rather would only fall back to `Connected` state.
On the Python side the user would interact with an object offering all the high level functionality and returning errors if they are called from an invalid state.

**Comment by pacman82** [comment]
* Approach 1 like requires upstream support from `odbc-api` allowing `StatementConnection` to own e.g. `Arc` connection types.
* Approach 2 requires some careful thinking about the schema member

**Comment by pacman82** [comment]
* Approach 1: Supporting `Arc` would likely imply relying more on the native thread safety of the drivers. At least if we want to keep concurrent fetching, we would need to also implement `Send` for `StatementConnection<Arc<_>>`. See also: https://docs.rs/odbc-api/latest/odbc_api/struct.StatementConnection.html#impl-Send-for-StatementConnection%3C'c%3E

* Approach 2: Schema member can likely be updated on state changes. I dislike that writer and reader would have a "merged" interface.

**Comment by pacman82** [comment]
Note to myself:

* Followed down approach 2. Yet this would lead to many repetition in the signatures of the Python interface. This would be cumbersome to maintain and unintuitive to the user.

Going forward we aim to combine the 1. and 2. I.e. We offer an explicit connection object. However internally it would be modeled as `Arc<Mutex<Option<Connection>>>` instead of `Arc<Mutex<Connection>>` allowing the reader or writer to take full ownership of the connection. The reusing of the connection would be implemented by giving it back to the original Arc pointer.

**Comment by timrburnham** [comment]
This sounds ideal! Would it be practical for the Python Connection object to directly offer an `execute()` method, for performing DDL or other 0-result statements? As you pointed out above, I was abusing `read_arrow_batches_from_odbc()` for side-effects, without reading any Arrow data.

**Comment by riziles** [comment]
@timrburnham , side question: I'm no expert on odbc, but wouldn't DuckDB -> ODBC -> Arrow be quite a bit slower then just DuckDB -> Arrow? Or is this just for testing?

**Comment by timrburnham** [comment]
@riziles I was just showing a test case for connection reuse. In reality, I'm using arrow-odbc to query non-Duck databases, usually batching the Arrow data into DuckDB for further processing, or PyArrow to write out Parquet files.

**Comment by pacman82** [comment]
@timrburnham FYI: I have your use-case still in mind. Some groundwork is laid. Yet I have currently a bottle neck in terms of available spare time.

**Comment by louiehernandez95** [comment]
Hello @pacman12  Recently began using your package for my own project and I am a fan. Was wondering how close things are for enabling connection reuse.

**Comment by pacman82** [comment]
@louiehernandez95 Sorry, for the delayed response. Hard to estimate. Currently progress is limited by my hard to predict capacity to work on things in my spare time. We may be able to identify smaller useful intermediate steps. Would you mind telling me, what the scenario is, for which you want to reuse connections?

**Comment by louiehernandez95** [comment]
@pacman82 no need to apologize I appreciate the follow up. After some time using your package I would actually say my use case is centered more around transaction control. A little bit of context, I am using this package to build out ETL jobs.

**Comment by pacman82** [comment]
@timrburnham `arrow-odbc 9.3.0` is released. It features the ability to explicitly instantiate connections in user code and to use and re-use them at your hearts content.

**Comment by hamstero** [comment]
Thanks for above explanation.

If I want to do concurrent inserts, or selects for that matter, using arrow-odbc-py and threads (imagine multiple writer threads, each working on separate tables), is there anything special I need to consider?

Normally pyodbc isn't thread-safe regarding connections, as far as I'm aware, so no reuse of connection across threads using that. How about arrow-odbc-py's connection handling?

And how does .enable_odbc_connection_pooling() fit into this? Can I just call a .connect in each thread, also having first called enable_odbc_connection_pooling, and trust that arrow-odbc-py and the Rust interface will do what's needed, so I don't have to think more about it? Even to a degree where inside a block writing loop I call .connect each time, because connection pooling is smart enough to reuse connection, so I don't have to manage it outside loop, but can be lazy and just call .connect anywhere I need it?

**Comment by pacman82** [comment]
> is there anything special I need to consider?

arrow-odbc is thread safe. The driver managers are thread safe. The ODBC standard demands that drivers are, too. Not all are though. One thing to consider there is a difference between being thread safe and actually being able to do multiple things at once. It might be safe to invoke a connection from multiple threads in parallel, but whether or not these queries are executed simultaneously or one after the other entirely depends on your driver. If you are not on Windows and use `unixODBC` it is worth checking the configured thread level. This is an Option in unixODBC to make non-thread safe driver thread safe by inserting mutices. Yet, of course, if the driver is already thread safe, it slows them down.

In practice setting `Threading = 0` in the driver configuration (i.e. trust the driver to be thread safe). And using one Connection per thread, would be my starting point, until I learn otherwise.

> And how does .enable_odbc_connection_pooling() fit into this?

Call it once in the beginning of your application, before creating any connection. It is actually the only function in arrow-odbc which is not thread safe, because it changes shared, global state.

> Even to a degree where inside a block writing loop I call .connect each time, because connection pooling is smart enough to reuse connection, so I don't have to manage it outside loop, but can be lazy and just call .connect anywhere I need it?

Connection pooling will reuse connections. That is the point. However, a writer is more than just a connection. It also binds buffers specific to the shape of the data you want to insert. So, if you insert something different every time you loose nothing, but if you stream from the same source and start from the connection each time, you repeat some of the overhead. You would still save the time needed to connect to the database. Overall, I would recommend using one batch writer to write the entire thing.

I do not know the shape of your problem or code, but just to throw something out there: Maybe do not loop yourself over batches. Just wrap the source into a Record Batch reader and throw that into the `insert_into_table`. With connection pooling enabled you would not even need to call connect manually.

Best, Markus

**Comment by hamstero** [comment]
I'm currently only on Windows, so can only test there, and don't know if Windows driver has a threading setting.

You said "It also binds buffers specific to the shape of the data you want to insert." which got me thinking more about the current situation. Before, using pyodbc, I had a prepared statement for reuse, and had a feeling I was missing something with arrow-odbc in the first (naive) setup to try it out. Whether it really impacts performance I don't yet really know / have confirmation/tests of.

Architecture is a set of readers, each a class instance running in own thread, passing (now) a pa RecordBatch via a conduit class to a Queue instance (back pressured) that writer threads can pick from. This so reading is concurrent with writing. A writer thread know of all possible target tables in can receive data for, and currently pulls a RecordBatch at a time and writes it, wrapping that single RecordBatch in an array and making a pa reader out of it for arrow-odbc-py to consume. So arrow-odbc-py isn't fed a continous reader. Not sure, yet, if I could either pass a reader reference, instead of RecordBatch, from reader thread to writer thread, or I could somehow make a continuous looking reader wrapper around the queue pulling...
Sometimes reader->writer is 1:1, other times 1:n (if unpacking hierarchical data to relational structure).

I will make some isolated tests when I find time so see if "starting from a connection" on each single RecordBatch vs a continous reader has a significant impact - just in a single/same thread to keep it simple. To see if possible repeated buffer binding is hurting.

**Comment by pacman82** [comment]
`arrow-odbc 10.4.0` Python bindings have just been released. This version features `Connection.execute`. It allows executing SQL Statements which do not return a cursor, or of which the applictation does not intend to read the cursor, without the workaround of relying on an empty reader. Together with the already released ability to share connections between readers and writers as well as manual transaction management, this is the last missing piece to solve the original user problem  as encounterd by @timrburnham.

@hamstero If you want to share your tests findings, feel free to open a new issue.
